src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs

```diff
@@ -4761,6 +4752,141 @@ void M()
+class C
+{
+    void M()
+    {
+        int i;
+        local$$(i);
+
+        void local(int j) { j++; M(); }
+    }
+}",
+                Captures($"\r\n{WorkspacesResources.Variables_captured_colon} this"));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)]
+        [WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")]
+        public async Task QuickInfoCapturesOnLambda()
+        {
+            await TestWithoutSpeculativeAsync(@"
+class C
+{
+    void M()
+    {
+        int i;
+        System.Action a = () =$$> { i++; M(); };
+    }
+}",
+                Captures($"\r\n{WorkspacesResources.Variables_captured_colon} this, i"));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)]
+        [WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")]
+        public async Task QuickInfoCapturesOnLambda2()
+        {
+            await TestAsync(@"
+class C
+{
+    void M()
+    {
+        int i;
+        System.Action<int> a = j =$$> { i++; j++; M(); };
+    }
+}",
+                Captures($"\r\n{WorkspacesResources.Variables_captured_colon} this, i"));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)]
+        [WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")]
+        public async Task QuickInfoCapturesOnLambda2_DifferentOrder()
+        {
+            await TestWithoutSpeculativeAsync(@"
+class C
+{
+    void M(int j)
+    {
+        int i;
+        System.Action a = () =$$> { M(); i++; j++; };
+    }
+}",
+                Captures($"\r\n{WorkspacesResources.Variables_captured_colon} this, j, i"));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)]
+        [WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")]
+        public async Task QuickInfoCapturesOnLambda3()
+        {
+            await TestWithoutSpeculativeAsync(@"
+class C
+{
+    void M()
+    {
+        int i;
+        int @this;
+        N(() =$$> { M(); @this++; }, () => { i++; });
+    }
+    void N(System.Action x, System.Action y) { }
+}",
+                Captures($"\r\n{WorkspacesResources.Variables_captured_colon} this, @this"));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)]
+        [WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")]
+        public async Task QuickInfoCapturesOnLambda4()
+        {
+            await TestWithoutSpeculativeAsync(@"
+class C
+{
+    void M()
+    {
+        int i;
+        N(() => { M(); }, () =$$> { i++; });
+    }
+    void N(System.Action x, System.Action y) { }
+}",
+                Captures($"\r\n{WorkspacesResources.Variables_captured_colon} i"));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)]
+        [WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")]
+        public async Task QuickInfoCapturesOnDelegate()
+        {
+            await TestWithoutSpeculativeAsync(@"
```

Comment:
Why would some tests test with speculative and some without it? We're only doing one thing in the quick info driver, right?

Reply:
The expected result when not using speculation is shown below, but the expected result when using speculation is empty.
I have not found a way (without extensive rewiring) to make a single test successfully check both.
So this test only checks the non-speculative case.
I can add test for the speculative case, but I'm not sure it's useful because it complicates the test machinery (need a tri-flag instead of bool) and it is trivial (empty result).

---
In reply to: [161964225](https://github.com/dotnet/roslyn/pull/23970#discussion_r161964225) [](ancestors = 161964225)

Reply:
if you're no longer having quick-info use a speculative model (which i havne't verified is what you're actually doing), is there any need for the quick-info tests to do anything special anymore?

Reply:
I've not been able to remove the usage of speculation yet. That's the main open issue that I'm working on.

Reply:
I removed the calls to `GetSemanticModelForNodeAsync` from the QuickInfo provider. 
The IDE now behaves as I'd expect (the captures are populated even after I edit the source).

Thanks for the tips! 😃